pilot/adapter/serviceregistry/aggregate/controller.go

```diff
@@ -15,167 +15,527 @@
 package aggregate
 
 import (
-	// TODO(nmittler): Remove this
-	_ "github.com/golang/glog"
-	multierror "github.com/hashicorp/go-multierror"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
 
 	"istio.io/istio/pilot/model"
 	"istio.io/istio/pilot/platform"
 	"istio.io/istio/pkg/log"
 )
 
+const (
+	// See labelsFromModel() for more on why we need
+	// internal / external representations
+	labelXdsPrefix = "config.istio.io/xds"
+
+	labelControllerPath = labelXdsPrefix + "ControllerPath"
+
+	labelServicePrefix = labelXdsPrefix + "Service."
+	// xDS external interface for host / service name
+	labelServiceName = labelServicePrefix + "name"
+	// xDS external interface for dns name
+	labelServiceExternalName = labelServicePrefix + "externalName"
+	// xDS external interface for service VIP
+	labelServiceVIP     = labelServicePrefix + "vip"
+	labelInstancePrefix = labelXdsPrefix + "ServiceInstance."
+	// xDS external interface for the instance IP.
+	// Exernal format is a valid ipv4/v6 string format, ex: 10.1.1.3
+	// Internally stored format is a hex representation
+	labelInstanceIP = labelInstancePrefix + "ip"
+	// xDS external interface for instance Port
+	// Exernal format is a valid 16 bit unsigned integer
+	// Internally stored format is a hex representation
+	labelInstancePort      = labelInstancePrefix + "port"
+	labelInstanceNamedPort = labelInstancePrefix + "namedPort"
+
+	// Value is Service Name + ip + port
+	labelInstanceRef = labelInstancePrefix + "reference"
+)
+
 // Registry specifies the collection of service registry related interfaces
 type Registry struct {
 	Name platform.ServiceRegistry
 	model.Controller
-	model.ServiceDiscovery
-	model.ServiceAccounts
+	// The service mesh view that this registry belongs to
+	MeshView *MeshResourceView
 }
 
-// Controller aggregates data across different registries and monitors for changes
-type Controller struct {
+// MeshResourceView is an aggregated store for resources sourced from various registries
+type MeshResourceView struct {
 	registries []Registry
+
+	// Mutex guards services, serviceInstances and corresponding labels
+	mu sync.RWMutex
+
+	// Canonical map of mesh service key to service references
+	services map[resourceKey]*model.Service
+
+	// Canonical map of mesh service instance keys to service instance references
+	serviceInstances map[resourceKey]*model.ServiceInstance
+
+	// A reverse map that associates label names to label values and their associated service resource keys
+	serviceLabels nameValueKeysMap
+
+	// A reverse map that associates label names to label values and their associated service instance resource keys
+	serviceInstanceLabels nameValueKeysMap
+
+	// Cache eviction handler for notifying changes to model objects
+	cacheEvictionHandler model.CacheEvictionHandler
```

Comment:
The backing store for this cache is the service registry and we poll/watch those registries. So, a cache eviction handler here makes little sense. Did you mean a changeNotificationCallback for the consumers?